DataFormats/Common/interface/getThinned_implementation.h

```diff
@@ -0,0 +1,171 @@
+#ifndef DataFormats_Common_getThinned_implementation_h
+#define DataFormats_Common_getThinned_implementation_h
+
+#include <optional>
+#include <tuple>
+
+#include "DataFormats/Common/interface/ThinnedAssociation.h"
+#include "DataFormats/Provenance/interface/BranchID.h"
+#include "DataFormats/Provenance/interface/ProductID.h"
+#include "DataFormats/Provenance/interface/ThinnedAssociationsHelper.h"
+
+namespace edm {
+  class WrapperBase;
+
+  namespace detail {
+    // This function provides a common implementation of
+    // EDProductGetter::getThinnedProduct() for EventPrincipal,
+    // DataGetterHelper, and BareRootProductGetter.
+    //
+    // getThinnedProduct assumes getIt was already called and failed to find
+    // the product. The input key is the index of the desired element in the
+    // container identified by ProductID (which cannot be found).
+    // If the return value is not null, then the desired element was
+    // found in a thinned container. If the desired element is not
+    // found, then an optional without a value is returned.
+    template <typename F1, typename F2, typename F3>
+    std::optional<std::tuple<WrapperBase const*, unsigned int> > getThinnedProduct(
+        ProductID const& pid,
+        unsigned int key,
+        ThinnedAssociationsHelper const& thinnedAssociationsHelper,
+        F1 pidToBid,
```

Comment:
You don't want to use `F1&&`?

Reply:
I actually had it first with `F1&&`, but then started to think otherwise https://github.com/cms-sw/cmssw/pull/30906#discussion_r460875735. The lambdas hold a pointer and possibly 64-bit integer, and the subsequent work would anyway need two forwarding calls, which would open risk for use-after-move (IIUC).